src/main/java/de/thomas_oster/liblasercut/drivers/Ruida.java

```diff
@@ -0,0 +1,1322 @@
+    /* start */
+    stream.hex("F10200");
+    /* light red */
+    stream.hex("D800");
+    /* feeding x, y */
+    stream.hex("E706").absoluteMM(job.getStartX()).absoluteMM(job.getStartY());
+
+    find_and_write_bounding_box(job);
+
+    int part_number = 0;
+
+    // layer count
+    stream.hex("CA22").byteint(job.getParts().size() - 1);
+
+    boolean first_prop = true;
+    boolean first_vector = true;
+
+    for (JobPart p : job.getParts())
+    {
+      float focus;
+
+      if ((p instanceof RasterPart) || (p instanceof Raster3dPart))
+      {
+        p = convertRasterizableToVectorPart((RasterizableJobPart)p, job, this.useBidirectionalRastering, true, true);
+      }
+      /* FALLTHRU */
+      if (p instanceof VectorPart)
+      {
+        double top_left_x = Util.px2mm(p.getMinX(), p.getDPI());
+        double top_left_y = Util.px2mm(p.getMinY(), p.getDPI());
+        double bottom_right_x = Util.px2mm(p.getMaxX(), p.getDPI());
+        double bottom_right_y = Util.px2mm(p.getMaxY(), p.getDPI());
+        /* write dimensions */
+        stream.hex("E752").byteint(part_number).absoluteMM(top_left_x).absoluteMM(top_left_y);
+        stream.hex("E753").byteint(part_number).absoluteMM(bottom_right_x).absoluteMM(bottom_right_y);
+        stream.hex("E761").byteint(part_number).absoluteMM(top_left_x).absoluteMM(top_left_y);
+        stream.hex("E762").byteint(part_number).absoluteMM(bottom_right_x).absoluteMM(bottom_right_y);  
+
+//        System.out.println("VectorPart(" + minX + ", " + minY + ", " + maxX + ", " + maxY + " @ " + p.getDPI() + "dpi)");
+
+        VectorPart vp = (VectorPart) p;
+          
+        //iterate over command list
+        for (VectorCommand cmd : vp.getCommandList())
+        {
+          //There are three types of commands: MOVETO, LINETO and SETPROPERTY
+          switch (cmd.getType())
+          {
+            case LINETO:
+            case MOVETO:
+            {
+              /**
+               * Move the laserhead (laser on) from the current position to the x/y position of this command.
+               */
+              if (first_vector) {
+                first_vector = false;
+
+                stream.hex("ca0100");
+                stream.hex("ca02").byteint(part_number); // start_layer
+                stream.hex("ca0113"); // blow on
+                stream.hex("c902").longint((int)speed);
+                // power for laser #1
+                stream.hex("c601").percent((int)min_power);
+                stream.hex("c602").percent((int)max_power);
+
+                /* start vector mode */
+                stream.hex("ca030f");
+                stream.hex("ca1000");
+              }
+              vector(cmd.getX(), cmd.getY(), p.getDPI(), cmd.getType() == CmdType.LINETO);
+              break;
+            }
+            case SETPROPERTY:
+            {
+              LaserProperty pr = cmd.getProperty();
+              if (pr instanceof FloatMinMaxPowerSpeedFrequencyProperty)
```

Comment:
If not, maybe print a warning? Can this happen?

Reply:
fixed.